pkg/controller/pubsub/pubsub_config_controller.go

```diff
@@ -137,9 +138,18 @@ func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
 	if err != nil {
 		return reconcile.Result{}, err
 	}
-	err = r.system.UpsertConnection(ctx, config, request.Name, cfg.Data["provider"])
-	if err != nil {
-		return reconcile.Result{}, err
+	if newConnFunc, ok := prvd.List()[cfg.Data["provider"]]; ok {
```

Comment:
Why is the retrieval of `newConnFunc` no longer wrapped by UpsertConnection()?

https://github.com/open-policy-agent/gatekeeper/blob/1878de1ce150142789cccb4be9ae30c44e3ad615/pkg/pubsub/system.go#L37-L70

Reply:
Two reasons - 
- We can determine in reconcile itself if `UpsertConnection` needs to be called or not by deciding if client config is suggesting opening a connection with an unsupported provider (felt this is a little better approach).
- I wasn't able to figure out how to write unit tests around `UpsertConnection` when it was wrapping retrieval of `newConnFunc`

Reply:
The problem with this is that it means that registered connections are no longer abstracted away by the connection system. We should not do this.

What was the difficulty in writing unit tests?

Reply:
To test that system's connection and providers map is properly being populated when initiating a new connection (ref - [system.go](https://github.com/open-policy-agent/gatekeeper/blob/master/pkg/pubsub/system.go#L58-L65)), I need to have [FakeNewConnection](https://github.com/open-policy-agent/gatekeeper/pull/2846/files#diff-5c7b335ef2fa2ff15f39349306a3ce266a7b2e88031eaab62644d82c43a3fd6cR368) constructors that provide fake instances that respect pubsub interface. If `UpsertConnection` wraps the retrieval of `newConnFunc`, then it returns the actual [NewConnection](https://github.com/open-policy-agent/gatekeeper/blob/master/pkg/pubsub/dapr/dapr.go#L63), and this function will return an error as it won't be able to establish an actual connection with running instances. There were two ways I could think of to cover the later half of the `system.UpserConnection` - 
1. Add fake providers in [this](https://github.com/open-policy-agent/gatekeeper/blob/master/pkg/pubsub/provider/provider.go#L10) list and then pass the provider name to call respective `FakeNewConnection`. This is dangerous as we do not want to allow users to mistakenly call fake providers.
2. Pass the `InitiateConnection` function as arg while calling `UpsertConnetion`. This made more sense.

I am open to ideas on how to handle this. 

I tried to be as unambiguous as I could be, but I know this explanation might be a little confusing. I can answer further questions or elaborate more if needed.

Reply:
Why not add a method to pkg/pubsub/provider/provider.go that lets tests inject providers at runtime?

Reply:
Or, alternatively, allow creation of a system with additional providers not recognized by pkg/pubsub/provider/provider.go .

This option is probably better, since the state will be automatically cleaned up, and wont interfere with other tests.

Reply:
Thanks for the suggestions! This didn't cross my mind. Reverted back to original code (wrapping provider retrieval in system).